ast_tools/stack.py

```diff
@@ -93,18 +93,26 @@ def inspect_enclosing_env(
         fn: tp.Callable, # tp.Callable[[tp.Dict[str, tp.Any], ...], tp.Any],
         *,
         decorators: tp.Optional[tp.Sequence[inspect.FrameInfo]] = None,
-        ) -> tp.Callable:
+        st: SymbolTable = None) -> tp.Callable:
     exec(_SKIP_FRAME_DEBUG_STMT)
     if decorators is None:
         decorators = ()
 
-    @functools.wraps(fn)
-    def wrapped_0(*args, **kwargs):
-        exec(_SKIP_FRAME_DEBUG_STMT)
-        st = get_symbol_table(list(itertools.chain(decorators, [wrapped_0])))
-        env = dict(st.globals)
-        env.update(st.locals)
-        return fn(env, *args, **kwargs)
-    return wrapped_0
+    if st is None:
+        @functools.wraps(fn)
+        def wrapped_0(*args, **kwargs):
+            exec(_SKIP_FRAME_DEBUG_STMT)
+            st = get_symbol_table(list(itertools.chain(decorators, [wrapped_0])))
+            env = dict(st.globals)
+            env.update(st.locals)
+            return fn(env, *args, **kwargs)
+    else:
+        @functools.wraps(fn)
+        def wrapped_0(*args, **kwargs):
+            exec(_SKIP_FRAME_DEBUG_STMT)
+            env = dict(st.globals)
+            env.update(st.locals)
+            return fn(env, *args, **kwargs)
```

Comment:
Can the `if` statement be moved inside the definition of wrapped so we don't have to duplicate the function definitions? IIRC there may be a gotcha with dynamic scoping, but from what i can tell, you could do something like

```
if st is None:
    st = get_symbol_table(list(itertools.chain(decorators, [wrapped_0])))
```

in my experience you may need to use functools.partial to make sure `st` isn't overwritten in the calling scope.

Reply:
That was what I originally tried to do, but it complains about possibly using an uninitialized var. Let me look into functools.partial

Reply:
@leonardt, fixed it to remove duplicate code